Main.hs

```diff
@@ -323,37 +323,52 @@ compile' def =
 
 compileInstance :: Definition -> TCM [Hs.Decl ()]
 compileInstance def = do
-  ih <- compileInstHead . unEl . defType $ def
+  ir <- compileInstRule [] (unEl . defType $ def)
   locals <- takeWhile (isAnonymousModuleName . qnameModule . fst)
           . dropWhile ((<= defName def) . fst)
           . sortDefs <$> curDefs
   ds <- mapM (compileInstanceClause locals) funClauses
-  return $
-    [Hs.InstDecl () Nothing (Hs.IRule () Nothing Nothing ih) (Just ds)]
+  return $ [Hs.InstDecl () Nothing ir (Just ds)]
   where Function{..} = theDef def
 
--- would be better to generate the whole InstDecl as the context also
--- needs to come from the type...
-compileInstHead :: Term -> TCM (Hs.InstHead ())
-compileInstHead ty = case unSpine $ ty of
-       -- only works with a completely concrete instance
-       Def f es | Just args <- allApplyElims es -> do
-         vs <- mapM (compileType . unArg) $ filter visible args
-         f <- hsQName f
-         return $ foldl (Hs.IHApp ()) (Hs.IHCon () f) vs
-       _ -> __IMPOSSIBLE__
-
+compileInstRule :: [Hs.Asst ()] -> Term -> TCM (Hs.InstRule ())
+compileInstRule cs ty = case unSpine  $ ty of
+  Def f es | Just args <- allApplyElims es -> do
+    vs <- mapM (compileType . unArg) $ filter visible args
+    f <- hsQName f
+    return $
+      Hs.IRule () Nothing (ctx cs) $ foldl (Hs.IHApp ()) (Hs.IHCon () f) (map pars vs)
+    where ctx [] = Nothing
+          ctx cs = Just (Hs.CxTuple () cs)
+          -- put parens around anything except a var or a constant
+          pars :: Hs.Type () -> Hs.Type ()
+          pars t@(Hs.TyVar () _) = t
+          pars t@(Hs.TyCon () _) = t
+          pars t = Hs.TyParen () t
```

Comment:
Is this needed? So far the pretty printer has been kind enough to put in parens where necessary.

Reply:
Without it I got `MonoidX a -> b`

Reply:
maybe it's a bug in the pretty printer for `Hs.IRule`?